js-packages/@prestojs/viewmodel/src/ViewModelCache.ts

```diff
@@ -674,12 +1095,157 @@ export default class ViewModelCache<
+                        return false;
+                    } catch (err) {
+                        if (err instanceof InvalidFieldError) {
+                            return err.message;
+                        }
+                        throw err;
+                    }
+                })
+                .filter(Boolean);
+            if (fieldErrors.length > 0) {
+                throw new Error(`Invalid field(s) provided: ${fieldErrors.join(', ')}`);
+            }
+        }
+
         const pkKey = this.getPkCacheKey(pk);
         const recordCache = this.cache.get(pkKey);
         if (!recordCache) {
             return null;
         }
-        return recordCache.get(fieldNames);
+        const relations: RelatedViewModelField[] = [];
+        const sourceFieldNames: FieldNames[] = [];
+        const nestedFields: Record<string, FieldPath[]> = {};
+        for (const pathElement of fieldNames) {
+            const fieldName: string = Array.isArray(pathElement)
+                ? pathElement[0]
+                : (pathElement as string);
+            const field = this.viewModel.fields[fieldName];
+            if (Array.isArray(pathElement)) {
+                const [name, ...p] = pathElement;
+                if (!nestedFields[name]) {
+                    nestedFields[name] = [];
+                }
+                if (p.length > 1) {
+                    // Nested fields - need to pass the whole array
+                    // eg. [user, group, id]
+                    nestedFields[name].push(p);
+                } else {
+                    // A specific field on this record
+                    // eg. [user, name]
+                    nestedFields[name].push(p[0]);
+                }
+            }
+            let sourceFieldName = fieldName as FieldNames;
+            if (field instanceof RelatedViewModelField) {
+                sourceFieldName = field.sourceFieldName as FieldNames;
+                if (!relations.includes(field)) {
+                    relations.push(field);
+                }
+                if (!Array.isArray(pathElement)) {
+                    if (!nestedFields[fieldName]) {
+                        nestedFields[fieldName] = [];
+                    }
+                    nestedFields[fieldName].push(
+                        ...field.to.fieldNames.filter(
+                            f =>
+                                !(field.to.fields[f] instanceof RelatedViewModelField) &&
+                                !nestedFields[fieldName].includes(f)
+                        )
+                    );
+                }
+            }
+            if (!sourceFieldNames.includes(sourceFieldName)) {
+                sourceFieldNames.push(sourceFieldName);
+            }
+        }
+        // If record exists under fieldNames key already then return it
+        let record = recordCache.get(fieldNames as FieldPath[]);
+        if (record) {
+            return record;
+        }
```

Comment:
Can you chat to me about this? I would have thought that you'd never store it with the related record and always look the record based on source field(s) only, and then inject the record reference back in

Reply:
easier implementation this way & easier to guarantee the record is always the same memory location so `===` always passes

Reply:
If someone updates the cache then doesn't it replace the record? (Or does it just update the fields on an existing record?)

If there's no test case for that already, can you make sure that replacing a record in the cache doesn't cause records that point to it to use the older (stale) record. ie pull back a User record with a related Group, then update the Group cache with a new Group record, then check that the User record correctly points to the new Group

Reply:
it replaces the record in the cache

yeah there's lots of tests around that